File: Leetcode/Python3/297. Serialize and Deserialize Binary Tree/solution.py
# https://leetcode.com/problems/serialize-and-deserialize-binary-tree/


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque


class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        queue = deque([root])
        answer = []
        while queue:
            node = queue.popleft()
            if node:
                answer.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                answer.append("#")

        print("".join(answer))
        return " ".join(answer)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if data == "#":  # If root is None serialize() return "#"
            return None
        nodes = data.split()

        root = TreeNode(nodes[0])
        queue = deque([root])
        index = 1
        while queue:
            node = queue.popleft()
            if nodes[index] != "#":
                node.left = TreeNode(nodes[index])
                queue.append(node.left)
            index += 1

            if nodes[index] != "#":
                node.right = TreeNode(nodes[index])
                queue.append(node.right)
            index += 1

        return root


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: Leetcode/C++/297. Serialize and Deserialize Binary Tree/solution.cpp
// https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

#include <queue>
#include <string>
#include <vector>
#include <sstream>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec
{
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode *root)
    {
        vector<string> encoded;
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty())
        {
            int length = q.size();
            for (int i = 0; i < length; i++)
            {
                TreeNode *cur = q.front();
                q.pop();
                if (cur == NULL)
                {
                    encoded.push_back("#");
                }
                else
                {
                    encoded.push_back(to_string(cur->val));

                    q.push(cur->left);
                    q.push(cur->right);
                }
            }
        }
        // Convert encoded vector to string
        stringstream ss;
        for (auto &s : encoded)
        {
            ss << s << " ";
        }
        return ss.str();
    }

    // Decodes your encoded data to tree.
    TreeNode *deserialize(string data)
    {
        // Except NULL TreeNode
        if (data == "# ")
        {
            return NULL;
        }
        // Split whitespace and save values to nodes vector
        vector<string> nodes;
        stringstream ss(data);
        string val;
        while (ss >> val)
        {
            nodes.push_back(val);
        }

        TreeNode *root = new TreeNode(stoi(nodes[0]));
        queue<TreeNode *> q;
        q.push(root);
        int index = 1;
        while (!q.empty())
        {
            TreeNode *cur = q.front();
            q.pop();
            // Connect left, right node
            if (nodes[index] != "#")
            {
                cur->left = new TreeNode(stoi(nodes[index]));
                q.push(cur->left);
            }
            index += 1;

            if (nodes[index] != "#")
            {
                cur->right = new TreeNode(stoi(nodes[index]));
                q.push(cur->right);
            }
            index += 1;
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));